### PR TITLE
Emit `delegate!` block errors

### LINKED ISSUE #395: Improve delegate! error handling
If you make any mistake in a `delegate` block, errors are swallowed.

The block expects a trait function to be declared, so without the `pub` keyword.
Any other syntax errors cause that all the delegated functions don't show up in the module.
The macro should emit `syn` errors to be presented in IDE.

```rust
delegate! {
        to self.erc20 {
            pub fn transfer(&mut self, recipient: &Address, amount: &U256); // pub
            fn transfer_from(&mut self, owner: &Address, recipient: &Address, amount: &U256);
            fn approve(&mut self, spender: &Address, amount: &U256);
            fn name(&self) -> String;
            fn symbol(&self) -> String;
            fn decimals(&self) - a-> u8; // wrong arrow
            fn total_supply(&self) -> U256;
            fn balance_of(&self, owner: &Address) -> U256;
            fn allowance(&self, owner: &Address, spender: &Address) -> U256;
        }

        to self.ownable {
            fn get_owner(&self) -> Address;
            fn transfer_ownership(&mut self, new_owner: &Address);
        }
    }
```

### PR DESCRIPTION
**Pull request by kpob**
## Summary by CodeRabbit

- **New Features**
	- Introduced error handling improvements for better resilience and stability.
- **Tests**
	- Added tests for invalid delegation scenarios.
- **Refactor**
	- Enhanced error handling in various components to improve code robustness.

### CODE REVIEW

**Comment by kubaplas** [comment]
Will it throw error when there's a typo in the code?

File: odra-macros/src/test_utils.rs [hunk 1]
```diff
@@ -101,6 +101,29 @@ pub mod mock {
...
 
+    pub fn module_invalid_delegation() -> ModuleImplIR {
+        let module = quote! {
+            impl Erc20 {
+                /// Returns the total supply of the token.
+                pub fn total_supply(&self) -> U256 {
+                    self.total_supply.get_or_default()
+                }
+
+                delegate! {
```
Comment (kubaplas) [inline odra-macros/src/test_utils.rs:112]: What is exactly invalid with this code?

Comment (kpob) [reply odra-macros/src/test_utils.rs:112]: the `pub` token in `pub fn get_owner(&self) -> Address;`

**Comment by kpob** [comment]
> Will it throw error when there's a typo in the code?

Yup, the place the error occurs will be highlighted in your IDE

File: odra-macros/src/test_utils.rs (hunk 1 above)
Comment (kubaplas) [reply odra-macros/src/test_utils.rs:112]: It cannot be handled in macro? 
Stan had a problem with hanging `>` somewhere in the delegate,  it gave no errors - for me this is more invalid.

Comment (kpob) [reply odra-macros/src/test_utils.rs:112]: It does not matter if is an extra `pub` or a hanging `>`, it is a syntax error that was suppressed and now is not. The macro expects `syn::TraitFn` or something like that. If cannot parse it, the error is propagated, and the whole `module` macro fails.

**Review by kubaplas — APPROVED** [review]

**Pull request merged**